lib/rack/query_parser.rb

```diff
@@ -98,7 +98,11 @@ def normalize_params(params, name, v, depth)
         params[k] ||= []
         raise ParameterTypeError, "expected Array (got #{params[k].class.name}) for param `#{k}'" unless params[k].is_a?(Array)
         params[k] << v
-      elsif after =~ %r(^\[\]\[([^\[\]]+)\]$) || after =~ %r(^\[\](.+)$)
+      elsif params[k].is_a?(Array) && after =~ %r(^\[(\d+)\]$)
+        index = $1.to_i
+        raise InvalidParameterError, "skipped Array index (got #{index} while current length is #{params[k].length}) for param `#{k}'" if index > params[k].length
```

Comment:
Done.

At first I inherited from `ArgumentError` which seemed to make sense but then I figured maybe `IndexError` makes more?

There was also a problem with using `ArgumentError`, it's too broadly caught. Shouldn't it be caught at the right times? (the current tests imply catching in `#unescape` and when matching `name` in `#normalize_params`).